### PR TITLE
Free port auto resolving for TarantoolServer and AppServer

### LINKED ISSUE #141: Fix 'Address already in use' error with TCP sockets
TcpPortDispatcher was introduced to eliminate the race condition when two workers trying to use the same port: the idea is that each worker use its own ports range. Really these ports could race with client ports (from, say, net.box or replication), which typically don't use bind() and so binds to a random available port (despite any dispatched ranges). So this approach is broken by design.

There is the workaround with `use_unix_sockets` and `use_unix_sockets_iproto` options: setting them to `True` should eliminate the problem. However it is possible that some tests are designed to test tcp connections, so I think we should fix the problem.

The way I see is no-bind / bind to zero port and grep logs for a LISTEN / CONSOLE port (or write it from a script to a predefined location). Remove TcpPortDispatcher and find_port() at all.

Hmm, `box.cfg{listen = 0}` shows `binary: bound to 0.0.0.0:0`. Need to elaborate how to get the port.

The discussion had initially arisen in #115.

### PR DESCRIPTION
**Pull request by ylobankov**
This change replaces manual choosing an iproto port for TarantoolServer
by the auto resolving mechanism. In two words, test-run always provides
'127.0.0.1:0' as a value for LISTEN env variable that is used in a lua
file to start a tarantool instance. In this way, the port will be picked
automatically, and we are getting the real value of it via the admin
console by executing `box.info.listen` that is available for tarantool
version >= 2.4.1, and special lua script intended for tarantool version
< 2.4.1.

The same for AppServer, test-run always provides '127.0.0.1:0' as a
value for LISTEN env variable that is used in a lua test script. In
this way, the iproto port will be picked automatically, but if the test
needs the real value of the port, it has to execute `box.info.listen`
for tarantool version >= 2.4.1, or other lua code for tarantool version
< 2.4.1.

Fixes #141

### CODE REVIEW

File: lib/tarantool_server.py [hunk 1]
```diff
@@ -906,8 +907,9 @@ def start(self, silent=True, wait=True, wait_load=True, rais=True, args=[],
         self.crash_detector.info = "Crash detector: %s" % self.process
         self.crash_detector.start()
 
+        deadline = time.time() + Options().args.server_start_timeout
```
Comment (Totktonada) [inline lib/tarantool_server.py:910]: Unnecessary change?

Comment (Totktonada) [reply lib/tarantool_server.py:910]: Ah, I see, `self.get_iproto_port(deadline)`. Sorry for noise.

File: lib/tarantool_server.py [hunk 2]
```diff
@@ -872,6 +868,11 @@ def start(self, silent=True, wait=True, wait_load=True, rais=True, args=[],
         color_log(prefix_each_line(' | ', self.version()) + '\n',
                   schema='version')
 
+        # When the server is restarted, we need to reset port.
+        if not self.use_unix_sockets_iproto:
+            if self.iproto.port > 0:
+                self._iproto = 0
```
Comment (Totktonada) [inline lib/tarantool_server.py:874]: TBH, this jugglish seems confusing for me:

```python
iproto = '127.0.0.1:0'
env['IPROTO'] = iproto
popen('tarantool')
iproto = get_iproto_port()  # '127.0.0.1:xxxx'
```

At different point the iproto variable means either an address to listen on (always with zero port) or an address to connect to (advertize address). I think the code would be simpler if we would separate those objects.

In fact, we don't need to store '127.0.0.1:0' anyhow, just place it to the env variable:

```python
env['IPROTO'] = '127.0.0.1:0'
open('tarantool')
iproto = get_iproto_address()
```

I don' insist, feel free to ignore.

File: lib/tarantool_server.py [hunk 3]
```diff
@@ -1270,3 +1276,79 @@ def wait_lsn(self, node_id, lsn):
...
+            # At first, it catches both server (listening) and client (sending)
+            # sockets (client sockets usually occur when starting an instance as
+            # a replica). Then it finds the server socket among caught sockets.
+            script = """
+                ffi = require('ffi')
+                socket = require('socket')
+
+                val = ffi.new('int[1]')
+                len = ffi.new('size_t[1]', ffi.sizeof('int'))
+                while not {deadline} or os.time() < {deadline} do
```
Comment (Totktonada) [inline lib/tarantool_server.py:1305]: When `deadline` is None, it'll be `while not None <...>` in Lua. It means accessing to a `None` variable. If it is undefined (it is likely so), this code will fail in the strict mode.

File: lib/tarantool_server.py [hunk 4]
```diff
@@ -1270,3 +1276,79 @@ def wait_lsn(self, node_id, lsn):
...
+                    for fd = 0, 65535 do
+                        local addrinfo = socket.internal.name(fd)
+                        local is_matched = addrinfo ~= nil and
+                            addrinfo.host == '{localhost}' and
+                            addrinfo.family == 'AF_INET' and
+                            addrinfo.type == 'SOCK_STREAM' and
+                            addrinfo.protocol == 'tcp' and
+                            type(addrinfo.port) == 'number'
+                        if is_matched then
+                            local key = socket.internal.SOL_SOCKET
```
Comment (Totktonada) [inline lib/tarantool_server.py:1315]: Nit: It is named `level` in `man getsockopt`.

File: lib/tarantool_server.py [hunk 5]
```diff
@@ -1270,3 +1276,79 @@ def wait_lsn(self, node_id, lsn):
...
+                                socket.internal.SO_OPT[key].SO_REUSEADDR.iname,
+                                val, len)
+                            if val[0] > 0 then
+                                return {{
+                                    host = addrinfo.host, port = addrinfo.port
+                                }}
+                            end
+                        end
+                    end
+                    os.execute("sleep " .. {sleep_time})
```
Comment (Totktonada) [inline lib/tarantool_server.py:1326]: Why not just `fiber.sleep({sleep_time})`? Moreover, if you stopped all the fibers, unlikely any of them will change anything. No reason to spin in the loop so.

File: lib/tarantool_server.py [hunk 6]
```diff
@@ -1270,3 +1276,79 @@ def wait_lsn(self, node_id, lsn):
...
+            return
+
+        deadline = deadline or 'nil'
+        sleep_time = 0.001
+        timeout = Options().args.server_start_timeout
+
+        version = yaml.safe_load(
+            self.admin("require('tarantool').version", silent=True)
+        )[0]
+        if version.split('.')[0] == '1':
```
Comment (Totktonada) [inline lib/tarantool_server.py:1293]: You can just execute this code if `box.info.listen` gives `nil`. No need explicit version check.

```lua
while <...> do
    local res = box.info.listen
    if res ~= nil then
        return <...>
    end
    -- <..tarantoool < 2.4.1 code..>
end
```


The explicit version check is worse than direct attempts to use a feature you need, at least because it'll work for versions before 2.4.0-103-g2c3acf2b0.

File: lib/tarantool_server.py [hunk 7]
```diff
@@ -1270,3 +1276,79 @@ def wait_lsn(self, node_id, lsn):
...
+                            type(addrinfo.port) == 'number'
+                        if is_matched then
+                            local key = socket.internal.SOL_SOCKET
+                            ffi.C.getsockopt(fd, key,
+                                socket.internal.SO_OPT[key].SO_REUSEADDR.iname,
+                                val, len)
+                            if val[0] > 0 then
+                                return {{
+                                    host = addrinfo.host, port = addrinfo.port
+                                }}
```
Comment (Totktonada) [inline lib/tarantool_server.py:1322]: Don't you want to detect the situation, when several listening ports are matched? It would be better to give an explicit error instead of returning a random address.

File: lib/app_server.py [hunk 8]
```diff
@@ -158,7 +158,7 @@ def deploy(self, vardir=None, silent=True, need_init=True):
             warn_unix_socket(path)
             self.iproto = path
         else:
-            self.iproto = str(find_port())
+            self.iproto = self.localhost + ':0'
```
Comment (Totktonada) [inline lib/app_server.py:161]: AFAIR it is only needed for the `create server` command with the `return_listen_uri` option. Now it'll return an unsuitable result.

Feel free to drop this feature together with the `self.iproto` field if you don't see a real need, but, please, don't leave it broken.

File: dispatcher.py [hunk 9]
```diff
@@ -46,35 +46,6 @@
 from listeners import StatisticsWatcher
 
 
-class TcpPortDispatcher:
```
Comment (Totktonada) [inline dispatcher.py:49]: How about removing of `utils.find_port()`?

**Review by Totktonada — APPROVED** [review]
Thank you for working on the ancient (but painful) problem!

I'm okay on the high level. I see a potential problem with the `create server` command, it should be handled somehow (see details in the comment above).

Also left several stylistic thoughts, but they're not crucial to fix.

I just looked at the code and lean on you in performing a thorough testing on tarantool and vshard test suites.

(I'm going to a vacation and can't participate in the next review iteration.)

File: lib/tarantool_server.py (hunk 2 above)
Comment (ylobankov) [reply lib/tarantool_server.py:874]: Agree, there is no sense to define `iproto` attribute until a real address is known.

File: lib/tarantool_server.py (hunk 3 above)
Comment (ylobankov) [reply lib/tarantool_server.py:1305]: I have above `deadline = deadline or 'nil'`.

File: lib/tarantool_server.py (hunk 6 above)
Comment (ylobankov) [reply lib/tarantool_server.py:1293]: We have some tests where a server starts without waiting for the full load. I don't know exactly, but I think there may be a situation when `box.info.listen` will give `nil` (that's why we do it in the loop) and in this case we will execute 'tarantoool < 2.4.1' code that may produce wrong result.

File: dispatcher.py (hunk 9 above)
Comment (ylobankov) [reply dispatcher.py:49]: `find_port()` is still used in inspector.py.

File: lib/app_server.py (hunk 8 above)
Comment (ylobankov) [reply lib/app_server.py:161]: No, it is not used for the `create server` command with the `return_listen_uri` option. The `create server` command uses `TarantoolServer` (please check [this](https://github.com/tarantool/test-run/blob/b59464f17f391ecffd985a2366777bc3cb354eca/lib/app_server.py#L54-L57)).

Actually, this is just an attribute that is used only [here](https://github.com/tarantool/test-run/blob/b59464f17f391ecffd985a2366777bc3cb354eca/lib/app_server.py#L36). I guess, the line `os.putenv("LISTEN", server.iproto)` exists only because some app tests have `box.cfg {listen = os.getenv('LISTEN')}` inside the test script. 

I can remove this attribute and move this `if ... else ...` block to the `run_server` function.

File: lib/tarantool_server.py [hunk 10]
```diff
@@ -774,7 +770,7 @@ def install(self, silent=True):
             warn_unix_socket(path)
             self._iproto = path
         else:
-            self._iproto = find_port()
+            self._iproto = 0
```
[the remark this answers is not in the document]
Comment (ylobankov) [reply lib/tarantool_server.py:773]: We can remove this line.

File: lib/app_server.py (hunk 8 above)
Comment (ylobankov) [reply lib/app_server.py:161]: Or rename it into `self.listen` to not be confused.

File: lib/tarantool_server.py (hunk 2 above)
Comment (ylobankov) [reply lib/tarantool_server.py]: Fixed.

File: lib/tarantool_server.py [hunk 11]
```diff
@@ -1270,3 +1276,79 @@ def wait_lsn(self, node_id, lsn):
...
+            # This is a simple script to get the server socket of the instance.
+            # At first, it catches both server (listening) and client (sending)
+            # sockets (client sockets usually occur when starting an instance as
+            # a replica). Then it finds the server socket among caught sockets.
+            script = """
+                ffi = require('ffi')
+                socket = require('socket')
+
+                val = ffi.new('int[1]')
+                len = ffi.new('size_t[1]', ffi.sizeof('int'))
```
[the remark this answers is not in the document]
Comment (ylobankov) [reply lib/tarantool_server.py]: Fixed.

File: lib/tarantool_server.py (hunk 4 above)
Comment (ylobankov) [reply lib/tarantool_server.py]: Fixed.

File: lib/tarantool_server.py (hunk 5 above)
Comment (ylobankov) [reply lib/tarantool_server.py]: Fixed.

File: lib/app_server.py (hunk 8 above)
Comment (ylobankov) [reply lib/app_server.py]: Fixed.

File: lib/tarantool_server.py (hunk 6 above)
Comment (Totktonada) [reply lib/tarantool_server.py:1289]: Doesn't it mean that `wait=False wait_load=False` now will perform waiting anyway?

File: dispatcher.py (hunk 9 above)
Comment (Totktonada) [reply dispatcher.py:49]: And there is the same race here, right? See, it is the simplest place: here we acquire a port, listen and use it later. Nothing prevent us from binding to zero, listen, get the real port and add it into the environment. Moreover, nothing actually prevent us from using Unix sockets here. A worker has one inspector, a worker has one working directory.

It would be interesting to reproduce the situation, when we get the port clash here. Maybe we observed it many times but didn't know the reason.

File: lib/tarantool_server.py (hunk 6 above)
Comment (Totktonada) [reply lib/tarantool_server.py:1289]: Doesn't it mean that we should either:

* if waiting is acquired, do it,
* otherwise run a greenlet to poll the server and set the `iproto` attribute (don't forget to finish it when the server is stopped)?

Comment (ylobankov) [reply lib/tarantool_server.py:1289]: > Doesn't it mean that `wait=False wait_load=False` now will perform waiting anyway?

Looks like not. Note, the `wait=False wait_load=False` and `wait=True wait_load=False` cases are identical from the point of view of getting the iproto socket. Look at [this](https://github.com/tarantool/test-run/blob/b59464f17f391ecffd985a2366777bc3cb354eca/lib/tarantool_server.py#L1143-L1152). When `wait=True wait_load=False`, we test only the admin connection and return. When  `wait=True wait_load=True`, `box.info.listen` gives the socket even though `box.info.status` is `loading`.

Comment (Totktonada) [reply lib/tarantool_server.py:1289]: You assume that `box.cfg()` is successful (or will be successful) and so binding a listening socket for iproto is successful. It may be not the case for a `wait_load=False` instance. AFAIR, we have test cases, where a supplementary server is expected to fail loading.

File: dispatcher.py (hunk 9 above)
Comment (ylobankov) [reply dispatcher.py:49]: Ok, it was trivial to switch to free port auto resolving in `TarantoolInspector`. Fixed.

Comment (Totktonada) [reply dispatcher.py:49]: Nice! (It would be good to mark the new two commits as part of/related/fixes the issue.)

File: lib/tarantool_server.py (hunk 6 above)
Comment (ylobankov) [reply lib/tarantool_server.py:1290]: I realized that we shouldn't get the iproto port if `wait=False` because we can get the port only when the admin connection can be established and tested (`wait=True` guarantees it). 

Also, I think we shouldn't get the iproto port when `crash_expected=True`. When a server is going to be failed due to some error (for example, when `box.cfg()` is unsuccessful) and we expect this, we have to provide `crash_expected=True`, otherwise the test will fail.

Comment (ylobankov) [reply lib/tarantool_server.py:1290]: I would like to get the iproto port only when `wait=True wait_load=True crash_expected=False`. In this case, we can use the code (without `while` loop) that you proposed in your first comment. It really simplifies the things, but if a server starts with `wait=False` or `wait_load=False`, the iproto socket will be missing for it. But it's not the case for tests where Unix sockets are used (almost all). Anyway, all tests passed when I did testing, and diff and app tests are deprecated in favor of luatest ones and not produced anymore.
